AlphaWallet/Core/Types/Tokens/AlphaWalletTokensService.swift

```diff
@@ -0,0 +1,195 @@
+//
+//  AlphaWalletTokensService.swift
+//  AlphaWallet
+//
+//  Created by Vladyslav Shepitko on 08.07.2022.
+//
+
+import Foundation
+import Combine
+import CombineExt
+
+class AlphaWalletTokensService: TokensService {
+    //FIXME: remove it later
+    private static let tokenUpdateBalanceQueue = DispatchQueue(label: "org.alphawallet.swift.tokenBalanceFetcher")
+    private var cancelable = Set<AnyCancellable>()
+    private var tokensChangedSubject: PassthroughSubject<Void, Never> = .init()
+    private let providers: CurrentValueSubject<ServerDictionary<TokenSourceProvider>, Never> = .init(.init())
+    private let autoDetectTransactedTokensQueue: OperationQueue = {
+        let queue = OperationQueue()
+        queue.name = "Auto-detect Transacted Tokens"
+        queue.maxConcurrentOperationCount = 1
+        return queue
+    }()
+    private let autoDetectTokensQueue: OperationQueue = {
+        let queue = OperationQueue()
+        queue.name = "Auto-detect Tokens"
+        queue.maxConcurrentOperationCount = 1
+        return queue
+    }()
+    private let sessionsProvider: SessionsProvider
+    private let analyticsCoordinator: AnalyticsCoordinator
+    private let importToken: ImportToken
+    private let tokensDataStore: TokensDataStore
+    private let transactionsStorage: TransactionDataStore
+    private let nftProvider: NFTProvider
+    private let assetDefinitionStore: AssetDefinitionStore
+
+    var objectWillChange: AnyPublisher<Void, Never> {
+        tokensChangedSubject.eraseToAnyPublisher()
+    }
+    private (set) var tokens: [Token] = []
+
+    var newTokens: AnyPublisher<[Token], Never> {
+        providers.map { $0.values }
+            .flatMapLatest { $0.map { $0.newTokens }.merge() }
+            .eraseToAnyPublisher()
+    }
+
+    init(sessionsProvider: SessionsProvider, tokensDataStore: TokensDataStore, analyticsCoordinator: AnalyticsCoordinator, importToken: ImportToken, transactionsStorage: TransactionDataStore, nftProvider: NFTProvider, assetDefinitionStore: AssetDefinitionStore) {
+        self.sessionsProvider = sessionsProvider
+        self.tokensDataStore = tokensDataStore
+        self.importToken = importToken
+        self.analyticsCoordinator = analyticsCoordinator
+        self.transactionsStorage = transactionsStorage
+        self.nftProvider = nftProvider
+        self.assetDefinitionStore = assetDefinitionStore
+    }
+
+    func tokens(for servers: [RPCServer]) -> [Token] {
+        return tokensDataStore.enabledTokens(for: servers)
+    }
+
+    func mark(token: TokenIdentifiable, isHidden: Bool) {
+        let primaryKey = TokenObject.generatePrimaryKey(fromContract: token.contractAddress, server: token.server)
+        tokensDataStore.updateToken(primaryKey: primaryKey, action: .isHidden(isHidden))
+    }
+
+    func token(for contract: AlphaWallet.Address) -> Token? {
+        //NOTE: same as below
+        return tokensDataStore.token(forContract: contract)
```

Comment:
@hboon what do you think about these two methods, do we need to get token directly from `tokensDataStore ` to route a call to right chain related provider?

Reply:
I'm not sure, what's the alternative? But either way, `tokens(for:)` should use the same mechanism, right?

```
func tokens(for servers: [RPCServer]) -> [Token] {
        return tokensDataStore.enabledTokens(for: servers)
    }
```

Reply:
yes

Reply:
Ok then. Do what looks right to you. We can tweak it later if necessary